File: ShopNowAndroid/app/src/androidTest/java/com/example/shopnow/ui/tests/ShopNowUITests.kt
package com.example.shopnow.ui.tests

import com.example.shopnow.MainActivity
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import org.junit.Rule
import org.junit.Test
import com.example.shopnow.ui.screens.LoginScreen
import com.example.shopnow.ui.screens.ProductsScreen
import com.example.shopnow.ui.screens.ProductDetailScreen
import com.example.shopnow.ui.screens.SettingsScreen

class ShopNowUiTests {

    @get:Rule
    val rule = createAndroidComposeRule<MainActivity>()

    private val validEmail = "test"
    private val validPassword = "radu"

    /**
     * 1) Task A: login -> loading visible -> list visible after delayed backend
     */
    @Test
    fun taskA_loginShowsLoading_thenDisplaysProductsList() {
        val login = LoginScreen(rule)
        val products = ProductsScreen(rule)

        login.assertDisplayed()
        login.login(validEmail, validPassword)

        products.assertLoadingVisible()
        products.waitForListVisible()
        products.assertDisplayed()
    }

    /**
     * 2) Favourites sync (detail -> list reflected)
     * Needs ProductTags.FAV_PREFIX + per-item fav tag in ProductList UI.
     */
    @Test
    fun favourites_toggleInDetail_reflectedInList() {
        val login = LoginScreen(rule)
        val products = ProductsScreen(rule)
        val detail = ProductDetailScreen(rule)

        login.login(validEmail, validPassword)
        products.waitForListVisible()

        products.openProduct("1")
        detail.assertDisplayed()
        detail.toggleFavorite()
        detail.back()

        products.assertItemFavoriteOn("1")
    }

    /**
     * 3) Logout resets session (returns to login)
     * Needs ProductTags.SETTINGS tag for the settings button on Products screen.
     */
    @Test
    fun settings_logout_resetsSession_andReturnsToLogin() {
        val login = LoginScreen(rule)
        val products = ProductsScreen(rule)
        val settings = SettingsScreen(rule)

        login.login(validEmail, validPassword)
        products.waitForListVisible()

        products.openSettings()
        settings.assertDisplayed()

        settings.logout()
        login.assertDisplayed()
    }

    /**
     * 4) Invalid login shows error and stays on login
     */
    @Test
    fun login_invalidCredentials_showsError() {
        val login = LoginScreen(rule)

        login.assertDisplayed()
        login.login("wrong", "wrong")
        login.assertErrorDisplayed()
        login.assertDisplayed()
    }

    /**
     * 5) Pull-to-refresh changes data (fake backend adds r2 on refresh)
     */
    @Test
    fun products_pullToRefresh_addsNewItem() {
        val login = LoginScreen(rule)
        val products = ProductsScreen(rule)

        login.login(validEmail, validPassword)
        products.waitForListVisible()

        products.pullToRefresh()

        // After refresh, FakeBackend should insert item with id "r2"
        products.waitForProductItem("r2")
    }
}

File: ShopNowIOS/ShopNowIOSUITests/ShopNow_UITests.swift
//
//  ShopNow.swift
//  ShopNowIOS
//
//  Created by Radu-Stefan Ranzascu on 06.01.2026.
//


import XCTest

final class ShopNow_UITests: XCTestCase {

    private let validEmail = "test@email.com"
    private let validPassword = "radu"

    private func launchApp() -> XCUIApplication {
        let app = XCUIApplication()
        app.launch()
        return app
    }

    // 1) Task A: login -> loading -> list visible after delay
    func test_taskA_loginShowsLoading_thenDisplaysProductsList() {
        let app = launchApp()
        let login = LoginScreen(app: app)
        let products = ProductsScreen(app: app)

        login.waitForDisplayed()
        login.login(email: validEmail, password: validPassword)
        products.waitForList(timeout: 12)
    }

    // 2) Favourites sync (toggle in detail -> reflected in list)
    func test_favourites_toggleInDetail_reflectedInList() {
        let app = launchApp()
        let login = LoginScreen(app: app)
        let products = ProductsScreen(app: app)
        let detail = ProductDetailScreen(app: app)

        login.login(email: validEmail, password: validPassword)
        products.waitForList()

        products.openProduct(id: "1")
        detail.toggleFavourite()
        detail.goBack()

        products.assertFavorite(id: "1", state: "on")
}

    // 3) Logout resets session and returns to login
    func test_settings_logout_resetsSession_andReturnsToLogin() {
        let app = launchApp()
        let login = LoginScreen(app: app)
        let products = ProductsScreen(app: app)
        let settings = SettingsScreen(app: app)

        login.login(email: validEmail, password: validPassword)
        products.waitForList()

        products.openSettingsScreen()
        settings.waitForDisplayed()

        settings.logoutUser()

        login.waitForDisplayed()
    }

    // 4) Invalid login shows error
    func test_login_invalidCredentials_showsError() {
        let app = launchApp()
        let login = LoginScreen(app: app)

        login.waitForDisplayed()
        login.login(email: "wrong", password: "wrong")

        login.assertErrorVisible()
    }

    // 5) Pull-to-refresh adds new item (assuming your iOS fake backend does the same)
    func test_products_pullToRefresh_addsNewItem() {
        let app = launchApp()
        let login = LoginScreen(app: app)
        let products = ProductsScreen(app: app)

        login.login(email: validEmail, password: validPassword)
        products.waitForList()

        products.pullToRefresh()
        products.assertProductExists(id: "r2", timeout: 10)
    }
}
